systems-analysis/webhook-based-automated-p2p-client-injection/README.md

```diff
@@ -0,0 +1,305 @@
+
+   ```yaml
+   apiVersion: v1
+   kind: Pod
+   metadata:
+     name: test-app-with-dfdaemon-socket
+     annotations:
+       dragonfly.io/inject-p2p: "true" # Annotation to trigger the Webhook
+   spec:
+     containers:
+       - name: test-app-container
+         image: test-app-image:latest
+         volumeMounts:
+           - name: dfdaemon-socket
+             mountPath: /var/run/dfdaemon.sock # Path to dfdaemon socket inside the container
+     volumes:
+       - name: dfdaemon-socket
+         hostPath:
+           path: /var/run/dfdaemon.sock # Actual path to dfdaemon socket on the node
+           type: Socket
+   ```
+
+3. **Binary Tool Injection**:
+   Considering that many base container images do not include the binary tool (such as `dfget`), and manual installation is inconvenient, this project will solve this problem using an Init Container. The Webhook will automatically add an initContainer to the target Pod. This initContainer is a custom lightweight image available in both amd64 and arm64 architectures, each containing the corresponding architecture's binary tool. The Webhook will also copy the binary tool from this initContainer to a shared volume. Subsequently, the Webhook modifies the `PATH` environment variable of the application container to add the shared volume directory where binary is located, allowing the application container to execute binary commands directly from the command line without additional user installation or specifying the full path.
+
+   The InitContainer uses Docker's manifest list to achieve the function of automatically importing the corresponding architecture initContainer, and its build commands are as follows:
+
+   ```bash
+   # Create manifest
+   docker manifest create dragonflyoss/d7y-p2p-injector:latest \
+   dragonflyoss/d7y-p2p-injector-amd64-linux:latest \
+   dragonflyoss/d7y-p2p-injector-arm64-linux:latest
+
+   docker manifest annotate dragonflyoss/d7y-p2p-injector-amd64-linux:latest --arch amd64 --os linux
+   docker manifest annotate dragonflyoss/d7y-p2p-injector-arm64-linux:latest --arch arm64 --os linux
+   docker manifest push dragonflyoss/d7y-p2p-injector:latest
+   ```
+
+   Sample yaml for the injected pod:
+
+   ```yaml
+   apiVersion: v1
+   kind: Pod
+   metadata:
+     name: test-app-with-dfget-init-image
+     annotations:
+       dragonfly.io/inject-p2p: "true" # Annotation to trigger the Webhook
+       dragonfly.io/d7y-p2p-injector-version: "2.2.3" # Specify the version of the injector image, default is latest if not specified
+   spec:
+     initContainers: # Injected by the webhook
+       - name: d7y-p2p-injector
+         image: dragonflyoss/d7y-p2p-injector:2.2.3
+         command: ["cp", "/usr/local/bin/dfget", "/dfget-tool/dfget"] # Copy dfget to the shared volume

```

Comment:
Is it possible to move this operations to the Dockerfile? to simplify the logic here.